ext/include/opentelemetry/ext/zpages/tracez_processor.h

```diff
@@ -0,0 +1,100 @@
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace ext
+{
+namespace zpages
+{
+/**
+ * The span processor passes finished recordables to an exporter and is
+ * used by the Data Aggregator to store spans before aggregating for TraceZ.
+ *
+ */
+class TracezSpanProcessor : public opentelemetry::sdk::trace::SpanProcessor {
+ public:
+  /**
+   * Initialize a span processor.
+   * @param exporter the exporter used by the span processor. Here, the purpose
+   * is still being discussed
+   */
+  explicit TracezSpanProcessor(std::unique_ptr<opentelemetry::sdk::trace::SpanExporter> &&exporter) noexcept
+      : exporter_(std::move(exporter)) {}
+
+  /**
+   * Create a span recordable using the associated exporter.
+   * @return a newly initialized recordable
+   */
+  std::unique_ptr<opentelemetry::sdk::trace::Recordable> MakeRecordable() noexcept override
+  {
+    return exporter_->MakeRecordable();
+  }
+
+  /**
+   * OnStart is called when a span starts; that span is added to running_spans.
+   * @param span a recordable for a span that was just started
+   */
+  void OnStart(opentelemetry::sdk::trace::Recordable &span) noexcept override;
+
+  /**
+   * OnEnd is called when a span ends; that span is moved from running_spans to
+   * completed_spans
+   * @param span a recordable for a span that was ended
+   */
+  void OnEnd(std::unique_ptr<opentelemetry::sdk::trace::Recordable> &&span) noexcept override;
+
+  /**
+   * Returns a reference to running_spans
+   * @return currently running spans when the function is called
+   */
+  std::unordered_set<opentelemetry::sdk::trace::Recordable*>& GetRunningSpans() noexcept;
+
+  /**
+   * Returns completed_spans, which is then cleared to make room for new
+   * completed spans
+   * @return newly completed spans, which is defined by spans that completed
+   * after the last call of this function
+   */
+  std::vector<std::unique_ptr<opentelemetry::sdk::trace::Recordable>> GetCompletedSpans() noexcept;
+
+  /**
+   * Export all ended spans that have not yet been exported.
+   * @param timeout an optional timeout, the default timeout of 0 means that no
+   * timeout is applied.
+   */
+  void ForceFlush(
+      std::chrono::microseconds timeout = std::chrono::microseconds(0)) noexcept override
+  {}
+
+  /**
+   * Shut down the processor and do any cleanup required. Ended spans are
+   * exported before shutdown. After the call to Shutdown, subsequent calls to
+   * OnStart, OnEnd, ForceFlush or Shutdown will return immediately without
+   * doing anything.
+   * @param timeout an optional timeout, the default timeout of 0 means that no
+   * timeout is applied.
+   */
+  void Shutdown(std::chrono::microseconds timeout = std::chrono::microseconds(0)) noexcept override
+  {
+    exporter_->Shutdown(timeout);
+  }
+
+ private:
+  std::unique_ptr<opentelemetry::sdk::trace::SpanExporter> exporter_;
+  std::unordered_set<opentelemetry::sdk::trace::Recordable*> running_spans_;
```

Comment:
Is the TraceZProcessor supposed to be a thread-safe class (do multiple threads share a single TraceZProcessor, or do they all have their own)?

Reply:
I think it's supposed to be thread-safe and hypothetical threads could share a single processor if I'm looking the spec correctly https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#interface-definition